sdk/framework/openapi.go

```diff
@@ -575,84 +565,188 @@ func specialPathMatch(path string, specialPaths []string) bool {
 // expandPattern expands a regex pattern by generating permutations of any optional parameters
 // and changing named parameters into their {openapi} equivalents.
 func expandPattern(pattern string) []string {
-	var paths []string
-
-	// Determine if the pattern starts with an alternation for multiple roots
-	// example (root1|root2)/(?P<name>regex) -> match['(root1|root2)/(?P<name>regex)','root1|root2','/(?P<name>regex)']
-	match := altRootsRe.FindStringSubmatch(pattern)
-	if len(match) == 3 {
-		var expandedRoots []string
-		for _, root := range strings.Split(match[1], "|") {
-			expandedRoots = append(expandedRoots, expandPattern(root+match[2])...)
-		}
-		return expandedRoots
+	// Happily, the Go regexp library exposes its underlying "parse to AST" functionality, so we can rely on that to do
+	// the hard work of interpreting the regexp syntax.
+	rx, err := syntax.Parse(pattern, syntax.Perl)
+	if err != nil {
+		// This should be impossible to reach, since regexps have previously been compiled with MustCompile in
+		// Backend.init.
+		panic(err)
```

Comment:
> Is there a possibility that the regex pattern scanned here does not come from `Backend.init`

The assurance comes from https://github.com/hashicorp/vault/blob/04b2461c91a86f4543d579c3895cdfdaf19387f8/sdk/framework/openapi.go#L239 explicitly passing in the value from the `Pattern` field of a `framework.Path` struct.

> or is not already compiled for some reason?

This is safe because you can't get to this code without being in a request handling context, meaning you've already traversed the start of `sdk/framework.(*Backend).HandleRequest` which has already called `sdk/framework.(*Backend).init` which has already panicked if the regexp was invalid:
https://github.com/hashicorp/vault/blob/04b2461c91a86f4543d579c3895cdfdaf19387f8/sdk/framework/backend.go#L472

Given earlier code will already have panicked, I don't think there's a need to write error handling logic that won't be invoked.

To my mind, this is a good application of the principal that panic is absolutely not appropriate for handling mistakes arising at runtime, but is appropriate when the programmer asserts that something should never happen, but still provides defined behaviour for when what shouldn't happen, does, due to a coding bug.

> I'm not sure whether Vault will just crash on a `panic` or recover in some way but it's probably not great regardless.

Since we're within the code flow of Go's `net/http` request handling, it traps the panic, logs a `http: panic serving` (or `http2: panic serving`) log message, and then closes the connection (HTTP/1) or returns an `INTERNAL_ERROR` status (HTTP/2).

The wider operation of the Vault server is not impacted, just the individual OpenAPI-related request fails.

Reply:
Thanks for the detailed explanation and context 👍